src/main/kotlin/com/cognifide/gradle/aem/pkg/DownloadTask.kt

```diff
@@ -0,0 +1,141 @@
+    final override val config = AemConfig.of(project)
+
+    @InputDirectory
+    private val prepareDir = AemTask.taskDir(project, PrepareTask.NAME)
+
+    @Internal
+    val props = PropertyParser(project)
+
+    @Input
+    val forceNewFlag = props.flag(FORCE_NEW_FLAG)
+
+    @Input
+    val extractFlag = props.flag(EXTRACT_FLAG)
+
+    private val checkoutConfig = CheckoutConfig(project, config)
+
+    @Input
+    var filterPathProp: String = props.string("aem.download.filterPath", "")
+
+    @Input
+    var filterRootsProp = props.list("aem.download.filterRoots")
+
+    private val checkoutFilter by lazy { checkoutConfig.determineFilter(filterRootsProp, filterPathProp) }
+
+    @Input
+    val instanceProp = props.string("aem.download.instance", "")
+
+    private val instance: Instance by lazy { checkoutConfig.determineInstance(instanceProp) }
+
+    init {
+        description = "Builds and downloads CRX package from remote instance"
+        group = AemTask.GROUP
+
+        baseName = config.packageName
+        isZip64 = true
+
+        //Empty package uploaded to aem will have 'shell' suffix
+        classifier = CLASSIFIER_SHELL
+
+        //Take only first filter.xml file from definition
+        duplicatesStrategy = DuplicatesStrategy.EXCLUDE
+
+        //Load zip configuration only when the task is in the graph
+        project.gradle.taskGraph.whenReady {
+            if (it.hasTask(this)) {
+                this.specShellPackage()
+            }
+        }
+
+        // Task is always executed when in the graph
+        outputs.upToDateWhen { false }
+    }
+
+    @TaskAction
+    override fun copy() {
+        //Prepare shell package
+        super.copy()
+
+        val sync = InstanceSync(project, instance)
+        logger.lifecycle("Uploading package $archivePath")
+        val packagePath = sync.uploadPackage(archivePath).path
+        logger.lifecycle("Building remote package $packagePath")
+        sync.buildPackage(packagePath)
+        logger.lifecycle("Downloading remote package $packagePath")
+        val downloaded = sync.downloadPackage(packagePath, destinationDir)
+
+        if(downloaded.exists()) {
+            if(extractFlag){
+                val jcrRoot = prepareJcrRoot()
+                extractContents(downloaded, jcrRoot)
+            }
+        } else {
+            throw FileException("Downloaded package missing: ${downloaded.path}")
+        }
+
+    }
+
+    private fun specShellPackage() {
+        into(PackagePlugin.VLT_PATH) { spec ->
+            spec.from(checkoutFilter.file)
+            rename(checkoutFilter.file.name, "filter.xml")
+        }
+        into("") { spec ->
+            spec.from(prepareDir)
+            spec.eachFile {
+                FileContentReader.filter(it) {
+                    //Updating version string for package to contain '-download' suffix
+                    props.expandPackage(it, mapOf("project.version" to "${project.version}-$CLASSIFIER_DOWNLOAD"), path)
```

Comment:
bad `path` usage. it is a project path but should be a path of file being expanded.

Reply:
There is possibility to break `aemDownload` because there could be defined custom `properties.xml` which will be copied by preparetask and that file could have hardcoded version instead of using var `{{project.version}}`.

You should rather postprocess entry in that ZIP rather than overriding that var. Please rethink it and redesign.

Reply:
Can't agree. The final package url is taken from the upload response, the operation will finish correctly, but in this case the downloaded package will be missing suffix and will have wrong version. 
In either case project team shouldn't hardcode project.version in the properties because it will cause released packages to be wrongly defined.

Reply:
it is at least problematic case, let's stay with what with got